### ISSUE TITLE
How to run Tensorpack training with deterministic behavior

### ISSUE THREAD

**Issue by jwook1004** [tensorpack/tensorpack#902]
Hi,

I tried to run [cifar10-resnet](https://github.com/tensorpack/tensorpack/blob/master/examples/ResNet/cifar10-resnet.py) training in Tensorpack/examples with the deterministic behavior, i.e., getting the same (or very similar) results for the repeated runs of the same setting.

Following the instructions in (https://github.com/tensorpack/tensorpack/issues/196), I made the following changes in my code:

Before [config](https://github.com/tensorpack/tensorpack/blob/2fc3be158ba5d6c002c8ae9f36afd7fbefb02a7d/examples/ResNet/cifar10-resnet.py#L160): 
`fix_rng_seed(seed)`
`tf.set_random_seed(seed)`

Inside [Model/optimizer](https://github.com/tensorpack/tensorpack/blob/2fc3be158ba5d6c002c8ae9f36afd7fbefb02a7d/examples/ResNet/cifar10-resnet.py#L114):
`opt = tf.train.MomentumOptimizer(lr, 0.9, use_locking=True)`

Inside [get_data()](https://github.com/tensorpack/tensorpack/blob/2fc3be158ba5d6c002c8ae9f36afd7fbefb02a7d/examples/ResNet/cifar10-resnet.py#L120):
`ds = PrefetchData(ds, 1, 1)`

In [TrainConfig()](https://github.com/tensorpack/tensorpack/blob/2fc3be158ba5d6c002c8ae9f36afd7fbefb02a7d/examples/ResNet/cifar10-resnet.py#L160), added session_config:
`session_config=get_default_sess_config(0.7)`
`session_config.intra_op_parallelism_threads = 1`
`session_config.inter_op_parallelism_threads = 1`

But, after all these changes, I can see noticeable difference in the results. More specifically, when I used one V100 gpu to run cifar10-resnet20 training three times for only 1 epoch, I got the following results:

Run1) train_error: 0.43832, validation_cost: 1.2135, validation_error: 0.437
Run2) train_error: 0.43001,  validation_cost: 1.6414, validation_error: 0.5324
Run3) train_error: 0.43894, validation_cost: 1.3848, validation_error: 0.4854

As you can see, these runs get slightly different train_errors, but quite different validation errors. This deviation in val-error is noticeable even after many epochs.

I checked that the above changes set the same values for the initial weight. So I conjecture that there's still some randomness coming from data fetch, especially when validation error is calculated, but not sure. Could you let me know if there's anything I missed from the above changes I made?

In https://github.com/tensorpack/tensorpack/issues/196, you mention that 
> After doing the above things I see very tiny differences when training mnist for a couple of epochs.

I was wondering how little the differences were... 

Best regards,
Jungwook

**Comment by ppwwyyxx** [comment]
I ran some tests with the `cifar-convnet.py` example. After those changes I'm not able to get deterministic results either, and observed similar gap in validation error.
However running the same code on CPU gave me deterministic results (up to the very last digit).
So I guess it's just how the ops work on GPU and there isn't much we can do in tensorpack.

**Comment by ppwwyyxx** [comment]
There are many related issues in tensorflow, e.g. https://github.com/tensorflow/tensorflow/issues/18096 . Closing here.

**Comment by duncanriach** [comment]
Hi @jwook1004, several sources of GPU-related non-determinism have now been addressed in TensorFlow. You may find that by setting the environment variable `TF_DETERMINISTIC_OPS` to "true" or "1" resolves the issues you were seeing.

Work is ongoing, but for more information, please see <https://github.com/NVIDIA/tensorflow-determinism>
